README.md
[ ![Codeship Status for Skalar/klarna-checkout-ruby](https://www.codeship.io/projects/c01f6250-8b2d-0131-29ef-223e3dbeac7e/status?branch=master)](https://www.codeship.io/projects/15721) [![Code Climate](https://codeclimate.com/github/Skalar/klarna-checkout-ruby.png)](https://codeclimate.com/github/Skalar/klarna-checkout-ruby) [![Gem Version](https://badge.fury.io/rb/klarna-checkout.png)](http://badge.fury.io/rb/klarna-checkout)

# Klarna Checkout

Unofficial Ruby Wrapper for Klarnas Checkout Rest API.

## Installation

Add this line to your application's Gemfile:

    gem 'klarna-checkout'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install klarna-checkout

## Usage

```ruby
require 'klarna/checkout'

# a) Create a client with shared_secret directly
client = Klarna::Checkout::Client.new({
  shared_secret: 'your-shared-secret',
  environment: :test # or :production
})

# b) Create a client with shared_secret from module configuration
Klarna::Checkout.configure do |config|
  config.shared_secret = 'your-shared-secret'
end

client = Klarna::Checkout::Client.new(environment: :test)
client.shared_secret
# => 'your-shared-secret'


# Initialize an order
order = Klarna::Checkout::Order.new({
  purchase_country: 'NO',
  purchase_currency: 'NOK',
  locale: 'nb-no',
  cart: {
    items: [{
      reference:  '1123581220325',
      name:       'Widget',
      quantity:   1,
      unit_price: 666,
      tax_rate:   2500
    }]
  },
  merchant: {
    id: '1337',
    terms_uri:        'http://www.example.com/terms',
    checkout_uri:     'http://www.example.com/checkout',
    confirmation_uri: 'http://www.example.com/confirmation_uri',
    push_uri:         'http://www.example.com/push'
  }
})


# Create the order with Klarna
client.create_order(order)
order.id
# => ID of the order (no other attributes are updated)


# Read an order from Klarna
order = client.read_order("1234ABCD")

# Update an existent order with cart items
order.update(cart: { items: [...] })

# Update an order with Klarna
order = client.update_order(order)

# Update the order status in Klarna
order = client.update_order(order, { status: 'created'})

# Configuring some global variables
Klarna::Checkout.configure do |config|
  config.shared_secret = 'your-shared-secret'
  config.merchant_id   = '12345'
  config.default_country  = 'NO'
  config.default_currency = 'NOK'
end


# Instead of repeating yourself with supplying the same attributes for each
# order you can configure some default attributes
Klarna::Checkout::Order.defaults = {
  purchase_country: 'NO',
  purchase_currency: 'NOK',
  locale: 'nb-no',
  merchant: {
    id: '1337',
    terms_uri:        'http://www.example.com/terms',
    checkout_uri:     'http://www.example.com/checkout',
    confirmation_uri: 'http://www.example.com/confirmation_uri',
    push_uri:         'http://www.example.com/push'
  }
}

# All newly created orders will then have these attributes
order = Klarna::Checkout::Order.new
order.purchase_country
# => 'NO'
```

# Klarna Order Management - XML-RPC API

```ruby
require 'klarna/api'

# Create a client with the merchant credentials
api_client = Klarna::Api::Client.new(
  shared_secret: 'your-shared-secret',
  merchant_id: '1337',
  environment: :test # or :production
)

# Activate the order reservation
api_client.activate('Reservation number')

# Activate the order reservation and send optional information
api_client.activate('Reservation number', { orderid1: 'my-order-id' })

```

## TODO

* Validation of objects according to documentation
* Respect readonly attributes

## Supported Rubies

* 1.9.2
* 1.9.3
* 2.0.0
* 2.1.0

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes, don't forget to add tests (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


CHANGELOG.md
## Version 1.1.5

* Adds support for Klarna XML-RPC API.
* Implements the 'Klarna Order Management' activate method.

## Version 1.1.4

* Allows specifying which Klarna order attributes to update. This can be used for only updating the order status or adding a merchant reference to an existent order.
* Allows updating an order after initializing it.
* Initial commit for 'Germany only' attributes support.
* Client update_order can also be used to create a new order if id is not present.

## Version 1.1.3

* Fix a bug where shipping address would not be passed to Klarna

## Version 1.1.2

* Fix a bug where GUI options were passed on persisted orders (not allowed)

## Version 1.1.1

* Fix a bug where GUI options would not be passed to Klarna

## Version 1.1.0

* Exceptions raised now include the response body

## Version 1.0.1

* Fix bug with assigning `[]` to `Cart#items`

## Version 1.0.0

* Inherit exceptions from StandardError
* Refactored `Client`-class
* Removed custom `deep_merge`-ext

## Version 0.0.6

* Client-side validations
* `Client#update_order`
* Global configuration and default values
* Fixes a bug with `Order#merchant_reference` format

## Version 0.0.5

* Exceptions when requests to server fail
* Cleanup in Gemspec
* CI

## Version 0.0.4

_Gone_

## Version 0.0.3

_Nothing_

## Version 0.0.2

* `Client#environment` attribute

## Version 0.0.1

* Bootstrapping
* `Client#create_order`
* `Client#read_order`


spec/lib/klarna/checkout/order_spec.rb
require 'spec_helper'

describe Klarna::Checkout::Order do
  describe ".initialize" do
    subject do
      described_class.new \
        id:     1,
        status: 'foo'
    end

    its(:id)     { should eq 1 }
    its(:status) { should eq 'foo' }
  end

  describe "attributes" do
    it { should have_attribute(:id) } #,                 :readonly)  }
    it { should have_attribute(:purchase_country) } #,   :mandatory) }
    it { should have_attribute(:purchase_currency) } #,  :mandatory) }
    it { should have_attribute(:locale) } #,             :mandatory) }
    it { should have_attribute(:status) } #,             :readonly)  }
    it { should have_attribute(:reference) } #,          :readonly)  }
    it { should have_attribute(:reservation) } #,        :readonly)  }
    it { should have_attribute(:started_at) } #,         :readonly)  }
    it { should have_attribute(:completed_at) } #,       :readonly)  }
    it { should have_attribute(:created_at) } #,         :readonly)  }
    it { should have_attribute(:last_modified_at) } #,   :readonly)  }
    it { should have_attribute(:expires_at) } #,         :readonly)  }
  end

  describe "relations" do
    it { should have_one(:merchant_reference, as: Klarna::Checkout::MerchantReference) }
    it { should have_one(:billing_address,    as: Klarna::Checkout::Address) }
    it { should have_one(:shipping_address,   as: Klarna::Checkout::Address) }
    it { should have_one(:cart,               as: Klarna::Checkout::Cart) }
    it { should have_one(:customer,           as: Klarna::Checkout::Customer) }
    it { should have_one(:merchant,           as: Klarna::Checkout::Merchant) }
    it { should have_one(:gui,                as: Klarna::Checkout::Gui) }
  end

  describe "validations" do
    subject(:valid_order) do
      described_class.new \
        merchant_reference: {
          orderid1: 'foo',
          orderid2: 'bar'
        },
        purchase_country:   'NO',
        purchase_currency:  'NOK',
        locale: 'nb-no',
        cart: {
          items: [{
            reference:  '1123581220325',
            name:       'Widget',
            quantity:   1,
            unit_price: 666,
            tax_rate:   2500
          }]
        },
        merchant: {
          id: '666666',
          terms_uri:        'http://www.example.com/terms',
          checkout_uri:     'http://www.example.com/checkout',
          confirmation_uri: 'http://www.example.com/confirmation_uri',
          push_uri:         'http://www.example.com/push'
        }
    end

    it { should be_valid }

    context "when status is 'checkout_incomplete'" do
      before(:each) do
        subject.status = 'checkout_incomplete'
      end

      it "is invalid without a purchase country" do
        subject.purchase_country = nil
        subject.should_not be_valid
      end

      it "is invalid without a purchase currency" do
        subject.purchase_currency = nil
        subject.should_not be_valid
      end

      it "is invalid without a locale" do
        subject.locale = nil
        subject.should_not be_valid
      end

      let(:cart_item) { subject.cart.items[0] }

      it "is invalid without a cart item reference" do
        cart_item.reference = nil
        subject.should_not be_valid
      end

      it "is invalid without a cart item name" do
        cart_item.name = nil
        subject.should_not be_valid
      end

      it "is invalid without a cart item quantity" do
        cart_item.quantity = nil
        subject.should_not be_valid
      end

      it "is invalid without a cart item unit price" do
        cart_item.unit_price = nil
        subject.should_not be_valid
      end

      it "is invalid without a cart item tax rate" do
        cart_item.tax_rate = nil
        subject.should_not be_valid
      end

      let(:merchant) { subject.merchant }

      it "is invalid without a merchant id" do
        merchant.id = nil
        subject.should_not be_valid
      end

      it "is invalid without a merchant terms_uri" do
        merchant.terms_uri = nil
        subject.should_not be_valid
      end

      it "is invalid without a merchant checkout_uri" do
        merchant.checkout_uri = nil
        subject.should_not be_valid
      end

      it "is invalid without a merchant confirmation_uri" do
        merchant.confirmation_uri = nil
        subject.should_not be_valid
      end

      it "is invalid without a merchant push_uri" do
        merchant.push_uri = nil
        subject.should_not be_valid
      end
    end

    context "when status is 'checkout_complete' or 'created'" do
      ['checkout_complete', 'created'].each do |status|
        before(:each) do
          subject.status = 'checkout_complete'
        end

      end
    end
  end

  describe "#as_json" do
    let(:order) do
      described_class.new \
        merchant_reference: {
          orderid1: 'foo',
          orderid2: 'bar'
        },
        purchase_country:   'NO',
        purchase_currency:  'NOK',
        locale: 'nb-no',
        cart: {
          items: [{
            reference:  '1123581220325',
            name:       'Widget',
            quantity:   1,
            unit_price: 666,
            tax_rate:   2500
          }]
        },
        gui: {
          layout: 'desktop'
        },
        merchant: {
          id: '666666',
          terms_uri:        'http://www.example.com/terms',
          checkout_uri:     'http://www.example.com/checkout',
          confirmation_uri: 'http://www.example.com/confirmation_uri',
          push_uri:         'http://www.example.com/push'
        },
        shipping_address: {
          street_address: 'Example Street 1',
          postal_code: '3045',
          city: 'Drammen',
          country: 'NO',
          email: 'test@example.com',
          phone: '99988777'
        }
    end

    let(:json_hash) { order.as_json }
    subject do
      json_hash
    end

    shared_examples_for "it's not included when order has ID" do
      context "when 'id' is set" do
        before(:each) do
          order.id = 'foobar'
        end

        it { should be_nil }
      end
    end

    describe "merchant_reference" do
      subject { json_hash[:merchant_reference] }

      its([:orderid1]) { 'foo' }
      its([:orderid2]) { 'bar' }
    end

    its([:purchase_country])   { should eq 'NO' }
    its([:purchase_currency])  { should eq 'NOK' }
    its([:locale])             { should eq 'nb-no' }

    describe "cart/items/0" do
      subject { json_hash[:cart][:items][0] }

      its([:reference])  { should eq '1123581220325' }
      its([:name])       { should eq 'Widget' }
      its([:quantity])   { should eq 1 }
      its([:unit_price]) { should eq 666 }
      its([:tax_rate])   { should eq 2500 }
    end

    describe "merchant" do
      subject { json_hash[:merchant] }

      its([:id])               { should eq '666666' }
      its([:terms_uri])        { should eq 'http://www.example.com/terms' }
      its([:checkout_uri])     { should eq 'http://www.example.com/checkout' }
      its([:confirmation_uri]) { should eq 'http://www.example.com/confirmation_uri' }
      its([:push_uri])         { should eq 'http://www.example.com/push' }

      it_behaves_like "it's not included when order has ID"
    end

    describe "gui" do
      subject { json_hash[:gui] }

      its([:layout]) { should eq 'desktop' }

      it_behaves_like "it's not included when order has ID"
    end

    describe "shipping_address" do
      subject { json_hash[:shipping_address] }

      its([:street_address]) { should eq 'Example Street 1' }
      its([:postal_code])    { should eq '3045' }
      its([:city])           { should eq 'Drammen' }
      its([:country])        { should eq 'NO' }
      its([:email])          { should eq 'test@example.com' }
      its([:phone])          { should eq '99988777' }
    end
  end

  describe "#to_json" do
    it "bases it output on #as_json" do
      subject.stub(:as_json) { { foo: "bar" } }
      subject.to_json.should eq JSON.generate({ foo: "bar" })

      subject.stub(:as_json) { { bar: "foo" } }
      subject.to_json.should eq JSON.generate({ bar: "foo" })
    end
  end

  describe ".defaults" do
    context "with some defaults set" do
      before(:each) do
        described_class.defaults = {
          purchase_country:   'NO',
          purchase_currency:  'NOK',
          merchant: {
            id: '666666'
          }
        }
      end
      after(:each) do
        described_class.defaults = nil
      end

      it "all new orders have the default values" do
        order = described_class.new
        order.purchase_country.should eq  'NO'
        order.purchase_currency.should eq 'NOK'
      end

      it "should be possible to override the default values" do
        order = described_class.new(purchase_currency: 'SEK')
        order.purchase_currency.should eq 'SEK'
      end

      it "should be possible to provide any nested values without affecting the defaults" do
        order = described_class.new({
          merchant: {
            terms_uri: 'http://www.example.com/terms'
          }
        })
        order.merchant.id.should eq '666666'
        order.merchant.terms_uri.should eq 'http://www.example.com/terms'
      end

      context "when I override an already specified default value" do
        before(:each) do
          described_class.defaults = {
            purchase_currency: 'SEK'
          }
        end

        it "newly created objects should use the new default value" do
          order = described_class.new
          order.purchase_currency.should eq 'SEK'
        end

        it "shouldn't remove the old default values" do
          order = described_class.new
          order.purchase_country.should eq  'NO'
          order.merchant.id.should eq '666666'
        end
      end
    end

    it "doesn't allow setting something other than a hash" do
      expect {
        described_class.defaults = :foobar
      }.to raise_error(ArgumentError)
    end
  end

  context "when Klarna::Checkout has been configured with some default values" do
    before(:each) do
      Klarna::Checkout.configure do |config|
        config.merchant_id = '424242'
        config.default_country  = 'NO'
        config.default_currency = 'NOK'
      end
    end

    after(:each) do
      Klarna::Checkout.reset_configuration!
    end

    it "should use the configuration's default values" do
      order = described_class.new
      order.purchase_country.should  eq 'NO'
      order.purchase_currency.should eq 'NOK'
      order.merchant.id.should       eq '424242'
    end

    context "if I specify another default merchant ID" do
      before(:each) do
        described_class.defaults = {
          purchase_country:  'SE',
          purchase_currency: 'SEK',
          merchant: {
            id: '666666'
          }
        }
      end

      it "should use my default values instead" do
        order = described_class.new
        order.purchase_country.should  eq 'SE'
        order.purchase_currency.should eq 'SEK'
        order.merchant.id.should       eq '666666'
      end
    end
  end

  describe "update" do
    context "with some values set" do
      let(:order) do
        described_class.new \
          merchant_reference: {
            orderid1: 'foo',
            orderid2: 'bar'
          },
          purchase_country:   'NO',
          purchase_currency:  'NOK',
          locale: 'nb-no',
          cart: {
            items: [
              {
                reference:  '1123581220325',
                name:       'Widget',
                quantity:   1,
                unit_price: 666,
                tax_rate:   2500
              },
              {
                reference:  '1123581220326',
                name:       'Widget 2',
                quantity:   3,
                unit_price: 222,
                tax_rate:   2500
              }
            ]
          },
          gui: {
            layout: 'desktop'
          },
          merchant: {
            id: '666666',
            terms_uri:        'http://www.example.com/terms',
            checkout_uri:     'http://www.example.com/checkout',
            confirmation_uri: 'http://www.example.com/confirmation_uri',
            push_uri:         'http://www.example.com/push'
          },
          shipping_address: {
            street_address: 'Example Street 1',
            postal_code: '3045',
            city: 'Drammen',
            country: 'NO',
            email: 'test@example.com',
            phone: '99988777'
          }
      end

      it 'should be possible to update some values without affecting the other values' do
        order.merchant.id.should eq '666666'
        order.merchant.terms_uri.should eq 'http://www.example.com/terms'
        order.merchant.update(id: '777777')
        order.merchant.id.should eq '777777'
        order.merchant.terms_uri.should eq 'http://www.example.com/terms'
        order.locale.should eq 'nb-no'

        order.cart.items.size.should eq 2

        item = order.cart.items.first
        item.reference.should eq '1123581220325'
        item.quantity.should eq 1

        order.update({
          cart: {
            items: [
              {
                reference:  '1123581220326',
                name:       'Widget 2',
                quantity:   4,
                unit_price: 223,
                tax_rate:   2500
              },
              {
                reference:  '1123581220327',
                name:       'Widget 3',
                quantity:   3,
                unit_price: 333,
                tax_rate:   2500
              },
              {
                reference:  '1123581220328',
                name:       'Widget 4',
                quantity:   1,
                unit_price: 444,
                tax_rate:   2500
              }
            ]
          }
        })

        order.locale.should eq 'nb-no'
        order.cart.items.size.should eq 3
        item = order.cart.items.first
        item.reference.should eq '1123581220326'
        item.quantity.should eq 4
        item.unit_price.should eq 223
      end
    end
  end

end


lib/klarna/api/client.rb
require 'xmlrpc/client'

module Klarna
  module Api
    # Klarna XML-RPC API Client
    class Client

      VERSION = 'ruby:api:1.1.3'
      KLARNA_API_VERSION = '4.1'
      VALID_ENVS = [:test, :production]

      attr_reader   :merchant_id, :environment
      attr_accessor :shared_secret, :keepalive, :country, :language, :currency

      def initialize(config = {})
        self.merchant_id = Klarna::Api.merchant_id
        self.shared_secret = Klarna::Api.shared_secret
        self.environment = Klarna::Api.environment || :test
        self.keepalive = false
        self.config = config
      end

      def merchant_id=(new_id)
        @merchant_id = new_id.to_i
      end

      def environment=(new_env)
        new_env = new_env.to_sym
        unless VALID_ENVS.include?(new_env)
          raise "Environment must be one of: #{VALID_ENVS.join(', ')}"
        end
        @environment = new_env
      end

      def config=(hash)
        hash.each do |key, value|
          setter = "#{key}="
          self.send(setter, value) if self.respond_to?(setter)
        end
      end

      # Activates a reservation matching the given reservation number.
      #
      # @see Klarna::Api::Activate
      #
      # @param [String] rno - Klarna order reservation number
      # @param [Hash] optional_info
      #
      # @return [Array] [risk_status, invoice_number]
      def activate(rno, optional_info = {})
        unless rno.present?
          raise 'Order reference must be present!'
        end

        defaults = {
          flags: Klarna::Api::Flags::RSRV_SEND_BY_EMAIL
        }

        params = Activate.params_list(self, rno, defaults.merge(optional_info))

        xmlrpc_client.call( 'activate', KLARNA_API_VERSION, VERSION, *params )
      end

      private

      # Initializes the XMLRPC::Client with klarna settings.
      #
      # @return [XMLRPC::Client] xmlrpc_client
      def xmlrpc_client
        return @xmlrpc_client if @xmlrpc_client

        @xmlrpc_client = XMLRPC::Client.new_from_hash(
          host: host,
          path: '/',
          port: 443,
          use_ssl: true
        )
        @xmlrpc_client.http_header_extra = headers
        @xmlrpc_client
      end

      # Returns the klarna host.
      #
      # @return [String] host
      def host
        if environment == :production
          'payment.klarna.com'
        else
          'payment.testdrive.klarna.com'
        end
      end

      # Returns the headers for the connection.
      #
      # @return [Hash] headers
      def headers
        @headers = {}
        @headers['Accept-Encoding'] = 'gzip,deflate'
        @headers['Accept-Charset']  = 'UTF-8,ISO-8859-1,US-ASCII'
        @headers['User-Agent']      = "XMLRPC::Client (Ruby #{RUBY_VERSION})"
        @headers['Content-Type']    = 'text/xml; charset=ISO-8859-1'
        @headers['Connection']      = 'close' unless keepalive
        @headers
      end

    end
  end
end

lib/klarna/api/activate.rb
require 'digest'

module Klarna
  module Api
    class Activate

      # Activates the order in klarna.
      #
      # optional_info structure
      #
      # Variable        Type      Description
      # -------------------------------------------------------------------------
      # orderid1        string    Order ID #1
      #
      # orderid2        string    Order ID #2
      #
      # flags           integer   4: Send the invoice by mail
      #                           8: Send the invoice by e-mail
      #                           512: "Klarna Mobil" transaction
      #
      # reference       string    The reference person for the purchase if it is
      #                           a company purchase. You can also use reference
      #                           variable to write a message or other important
      #                           information to the consumer on the invoice.
      #
      # reference_code  string    The reference code for the sale. You can also
      #                           use reference_code variable to write a message
      #                           or other important information to the consumer
      #                           on the invoice.
      #
      # ocr             string    The payment reference number which the consumer
      #                           will use to pay Klarna. Please ask for more
      #                           information from your integration sales contact.
      #
      # pin             string    Consumer’s PIN code when using “Klarna Mobil”
      #                           service
      #
      # artnos          hash      Used if you want to activate a part of an order.
      #                           See Artnos structure below for more information.
      #
      # shipping_info   hash      Used to send information about shipment like
      #                           tracking numbers. See shipment_info structure
      #                           below for more information.
      #
      #
      # artnos structure
      #
      # Variable  Type      Description
      # -----------------------------------------------------------
      # artno     string    Article number on item to be activated
      # qty       integer   Number of items to activate
      #
      #
      # TODO: shipment_info structure
      #
      # @see https://developers.klarna.com/en/api-references-v1/manage-orders#activate
      #
      # @param [Klarna::Api::Client] client
      # @param [String] rno - Klarna order reservation number
      # @param [Hash] optional_info
      #
      #
      # @return [Array] [risk_status, invoice_number]
      #
      # risk_status - This represents the risk status and can have two values:
      #
      #   OK -      this response means that the order has passed Klarna’s fraud
      #              and credit assessment
      #   no_risk - this response means that Klarna will not assume the fraud
      #             risk for this order
      #
      # invoice_number - This number represents the invoice number in Klarna’s
      #                  system and should be used in all subsequent order
      #                  handling calls.
      #
      #
      # @raise [XMLRPC::FaultException] FaultException has two accessor-methods
      #                                 faultCode an Integer, and faultString a
      #                                 String.
      #
      def self.params_list(client, rno, optional_info = {})
        [
          client.merchant_id,
          digest(client, rno, optional_info),
          rno,
          optional_info
        ]
      end

      private

      # Digest
      #
      # Build digest. Any field in optional_info that is set is included in
      # the digest.
      #
      # Consist of all values from the variables which you send in the call
      # separated with a colon without the square brackets. Variables in
      # parentheses are optional.
      # If you are not sending a variable, remove that from the digest. Shared
      # secret must be the last variable.
      #
      # Example:
      #   base64encode(sha512("4:1:[client_vsn]:[eid]:[rno]:[flags](:[orderid1])
      #         (:[orderid2])(:[referece])(:[reference_code]):[shared_secret]"))
      #
      # @param [Klarna::Api::Client] client
      # @param [String] rno
      # @param [Hash] optional_info
      #
      # @return [String] digest
      def self.digest(client, rno, optional_info)

        optional_keys = [
          :bclass,
          :cust_no,
          :flags,
          :ocr,
          :orderid1,
          :orderid2,
          :reference,
          :reference_code
        ]

        digest_optional_info = optional_info.values_at(*optional_keys).compact

        if optional_info[:artnos]
          optional_info[:artnos].each do |article|
            digest_optional_info.push article[:artno]
            digest_optional_info.push article[:qty]
          end
        end

        array = [
          client.class::KLARNA_API_VERSION.gsub('.', ':'),
          client.class::VERSION,
          client.merchant_id,
          rno,
          *digest_optional_info,
          client.shared_secret
        ]

        ::Digest::SHA512.base64digest(array.join(':'))
      end

    end
  end
end

lib/klarna/checkout/resource.rb
require 'json'
require 'active_support/core_ext/hash/slice'
require 'active_support/core_ext/hash/deep_merge'

require 'klarna/checkout/concerns/has_one'
require 'klarna/checkout/concerns/has_many'

module Klarna
  module Checkout
    class Resource
      extend HasOne
      extend HasMany

      def initialize(args = {})
        self.class.defaults.deep_merge(args).each_pair do |attr, value|
          setter = "#{attr.to_s}="
          self.send(setter, value) if respond_to?(setter)
        end
      end

      def update(args = {})
        args.each_pair do |attr, value|
          setter = "#{attr.to_s}="
          self.send(setter, value) if respond_to?(setter)
        end
        self
      end

      def to_json(*keys)
        sanitized_json = json_sanitize(self.as_json, keys)
        JSON.generate(sanitized_json)
      end

      def json_sanitize(hash, keys = [])
        hash.reject! { |k, v| v.nil? }
        hash.slice!(*Array(keys)) if keys.any?
        hash
      end

      class << self
        def defaults=(hash)
          if hash
            raise ArgumentError.new unless hash.is_a? Hash

            @defaults ||= {}
            @defaults.deep_merge!(hash)
          else
            @defaults = {}
          end
        end

        def defaults
          @defaults || {}
        end
      end
    end
  end
end


lib/klarna/api.rb
require 'xmlrpc/client'
require 'digest'

module Klarna

  # Klarna XML-RPC API
  class Api

    VERSION = 'ruby:api:1.1.3'

    KLARNA_API_VERSION = '4.1'

    attr_accessor :keepalive, :merchant_id, :secret, :country, :language,
      :currency, :server, :pc_storage, :pc_uri, :environment, :use_ssl, :host,
      :path, :port

    def initialize(config = {})
      @path = '/'
      @port = 443
      @use_ssl = true
      @keepalive = false
      @environment = :production
      @pc_storage = 'json'
      @pc_uri = './pclasses.json'
      self.config = config
    end

    def config=(hash)
      hash.each do |key, value|
        setter = "#{key}="
        self.send(setter, value) if self.respond_to?(setter)
      end
    end

    # Activates the order in klarna.
    #
    # optional_info structure
    #
    # Variable        Type      Description
    # -------------------------------------------------------------------------
    # orderid1        string    Order ID #1
    #
    # orderid2        string    Order ID #2
    #
    # flags           integer   4: Send the invoice by mail
    #                           8: Send the invoice by e-mail
    #                           512: "Klarna Mobil" transaction
    #
    # reference       string    The reference person for the purchase if it is
    #                           a company purchase. You can also use reference
    #                           variable to write a message or other important
    #                           information to the consumer on the invoice.
    #
    # reference_code  string    The reference code for the sale. You can also
    #                           use reference_code variable to write a message
    #                           or other important information to the consumer
    #                           on the invoice.
    #
    # ocr             string    The payment reference number which the consumer
    #                           will use to pay Klarna. Please ask for more
    #                           information from your integration sales contact.
    #
    # pin             string    Consumer’s PIN code when using “Klarna Mobil”
    #                           service
    #
    # artnos          hash      Used if you want to activate a part of an order.
    #                           See Artnos structure below for more information.
    #
    # shipping_info   hash      Used to send information about shipment like
    #                           tracking numbers. See shipment_info structure
    #                           below for more information.
    #
    #
    # artnos structure
    #
    # Variable  Type      Description
    # -----------------------------------------------------------
    # artno     string    Article number on item to be activated
    # qty       integer   Number of items to activate
    #
    #
    # TODO: shipment_info structure
    #
    # @see https://developers.klarna.com/en/api-references-v1/manage-orders#activate
    #
    # @param [String] rno - Klarna order reservation number
    # @param [Hash] optional_info
    #
    #
    # @return [Array] [risk_status, invoice_number]
    #
    # risk_status - This represents the risk status and can have two values:
    #
    #   OK -      this response means that the order has passed Klarna’s fraud
    #              and credit assessment
    #   no_risk - this response means that Klarna will not assume the fraud
    #             risk for this order
    #
    # invoice_number - This number represents the invoice number in Klarna’s
    #                  system and should be used in all subsequent order
    #                  handling calls.
    #
    #
    # @raise [XMLRPC::FaultException] FaultException has two accessor-methods
    #                                 faultCode an Integer, and faultString a
    #                                 String.
    #
    def activate(rno, optional_info = {})
      params_list = [
        merchant_id.to_i,
        digest(rno, optional_info),
        rno,
        optional_info
      ]

      xmlrpc_client.call('activate', KLARNA_API_VERSION, VERSION, *params_list)
    end

    # Digest
    #
    # Build digest. Any field in optional_info that is set is included in
    # the digest.
    #
    # Consist of all values from the variables which you send in the call
    # separated with a colon without the square brackets. Variables in
    # parentheses are optional.
    # If you are not sending a variable, remove that from the digest. Shared
    # secret must be the last variable.
    #
    # Example:
    #   base64encode(sha512("4:1:[client_vsn]:[eid]:[rno]:[flags](:[orderid1])
    #         (:[orderid2])(:[referece])(:[reference_code]):[shared_secret]"))
    #
    # @param [String] rno
    # @param [Hash] optional_info
    #
    def digest(rno, optional_info)

      optional_keys = [
        :bclass,
        :cust_no,
        :flags,
        :ocr,
        :orderid1,
        :orderid2,
        :reference,
        :reference_code
      ]

      digest_optional_info = optional_info.values_at(*optional_keys).compact

      if optional_info[:artnos]
        optional_info[:artnos].each do |article|
          digest_optional_info.push article[:artno]
          digest_optional_info.push article[:qty]
        end
      end

      array = [
        KLARNA_API_VERSION.gsub('.', ':'),
        VERSION,
        merchant_id.to_i,
        rno,
        *digest_optional_info,
        secret
      ]

      ::Digest::SHA512.base64digest(array.join(':'))
    end

    # Initializes the connection to klarna.
    #
    # @return [XMLRPC::Client] xmlrpc_client
    def xmlrpc_client
      return @xmlrpc_client if @xmlrpc_client

      @xmlrpc_client = XMLRPC::Client.new_from_hash(
        host: host,
        path: path,
        port: port,
        use_ssl: use_ssl
      )
      @xmlrpc_client.http_header_extra = headers
      @xmlrpc_client
    end

    # Returns the klarna host.
    #
    # @return [String] host
    def host
      if @host
        @host
      elsif environment == :production
        'payment.klarna.com'
      else
        'payment.testdrive.klarna.com'
      end
    end

    private

    # Returns the headers for the connection.
    #
    # @return [Hash] headers
    def headers
      @headers = {}
      @headers['Accept-Encoding'] = 'gzip,deflate'
      @headers['Accept-Charset']  = 'UTF-8,ISO-8859-1,US-ASCII'
      @headers['User-Agent']      = "XMLRPC::Client (Ruby #{RUBY_VERSION})"
      @headers['Content-Type']    = 'text/xml; charset=ISO-8859-1'
      @headers['Connection']      = 'close' unless keepalive
      @headers
    end

  end
end
